api/types/src/transaction.rs

```diff
@@ -1,168 +1,286 @@
+            gas_unit_price: txn.gas_unit_price().into(),
+            gas_currency_code: txn.gas_currency_code().to_owned(),
+            expiration_timestamp_secs: txn.expiration_timestamp_secs().into(),
+            events,
+            payload,
+        }))
+    }
+}
+
+impl<T: TransactionInfoTrait> From<(u64, &T, WriteSetPayload, Vec<Event>)> for Transaction {
+    fn from((version, info, payload, events): (u64, &T, WriteSetPayload, Vec<Event>)) -> Self {
+        Transaction::GenesisTransaction(GenesisTransaction {
+            version: version.into(),
+            hash: info.transaction_hash().into(),
+            state_root_hash: info.state_root_hash().into(),
+            event_root_hash: info.event_root_hash().into(),
+            gas_used: info.gas_used().into(),
+            success: info.status().is_success(),
+
+            payload: GenesisPayload::WriteSetPayload(payload),
+            events,
+        })
     }
 }
 
+impl<T: TransactionInfoTrait> From<(u64, &BlockMetadata, &T)> for Transaction {
+    fn from((version, txn, info): (u64, &BlockMetadata, &T)) -> Self {
+        Transaction::BlockMetadataTransaction(BlockMetadataTransaction {
+            version: version.into(),
+            hash: info.transaction_hash().into(),
+            state_root_hash: info.state_root_hash().into(),
+            event_root_hash: info.event_root_hash().into(),
+            gas_used: info.gas_used().into(),
+            success: info.status().is_success(),
+
+            id: txn.id().into(),
+            round: txn.round().into(),
+            previous_block_votes: txn
+                .previous_block_votes()
+                .clone()
+                .iter()
+                .map(|a| (*a).into())
+                .collect(),
+            proposer: txn.proposer().into(),
+        })
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct PendingTransaction {
+    pub hash: HashValue,
+    pub sender: Address,
+    pub sequence_number: U64,
+    pub max_gas_amount: U64,
+    pub gas_unit_price: U64,
+    pub gas_currency_code: String,
+    pub expiration_timestamp_secs: U64,
+    pub payload: TransactionPayload,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct UserTransaction {
+    pub version: U64,
+    pub hash: HashValue,
+    pub state_root_hash: HashValue,
+    pub event_root_hash: HashValue,
+    pub gas_used: U64,
+    pub success: bool,
+
+    // user txn specific fields
+    pub sender: Address,
+    pub sequence_number: U64,
+    pub max_gas_amount: U64,
+    pub gas_unit_price: U64,
+    pub gas_currency_code: String,
+    pub expiration_timestamp_secs: U64,
+    pub payload: TransactionPayload,
+    pub events: Vec<Event>,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct GenesisTransaction {
+    pub version: U64,
+    pub hash: HashValue,
+    pub state_root_hash: HashValue,
+    pub event_root_hash: HashValue,
+    pub gas_used: U64,
+    pub success: bool,
```

Comment:
have you considered

Transaction {
  // common_fields
  transaction_data: TransactionData
}

where the TransactionData can be these unique fields around an enum? it feels bad to have so much copy pasta, but I imagine there's some other ergonomics you're aiming for?

Reply:
From reader's point view of the rendered data, I think:
```
{ version: xxx, hash: xxx, ... }
```
is better than 
```
{ data: {version: xxx, hash: xxx}, ... }
```
fields like version and hash maybe used frequently, adding one level to the data structure feels odd to user who does not know why / how internal transaction data is structured.
I don't think a user is required to know how we organize transaction data internally for using the API, the goal of this API is hide unnecessary internal complexity for API user.

Technically, we can use optional fields to unify same fields
```
struct Transaction{
    version, 
    hash,
    payload: Option<TransactionPayload>,
    .....
}
```
or, use a customized serializer, to flat the transaction_data into the root Transaction object,
But these optionals creates unnecessary complexity for how we serialize data into JSON.

Thus we create the rust types that matches our internal data structure, but matches output data structure by serde configuration.

Overall, my principle is find the best balance between 2 data structures (because we need convert data between them, and we need control the complexity of serializer and deserializer):
1. Our Rust internal data structure: which is structured, opinionated and complex, learning curve is high.
2. API data structure that is aiming for easier to be interpreter by client applications in various languages without deep knowledge of Diem core data types. In general flat data structure is easier to understand comparing with deep hierarchy data structure. One reason is because a client application is more likely to only use partial data from the response, thus less hierarchy data structure may serve it better for ignoring unrelated data.

Reply:
alright, found serde flatten option to share fields by a struct, updated.